### PR TITLE
[red-knot] Add `walk_directories` to `System`

### PR DESCRIPTION
**Pull request by MichaReiser**
## Summary

This PR adds a `ignore::WalkDir` like API to `System` that allows walking a directory tree. 

We'll need this to implement the analysis of an entire directory. 

## Test Plan

Added unit tests.

### CODE REVIEW

File: crates/ruff_db/src/system/memory_fs.rs [hunk 1]
```diff
@@ -238,32 +253,30 @@ impl MemoryFileSystem {
         normalized.into_utf8_path_buf()
     }
 
-    pub fn read_directory(
-        &self,
-        path: impl AsRef<SystemPath>,
-    ) -> Result<impl Iterator<Item = Result<DirectoryEntry>> + '_> {
+    pub fn read_directory(&self, path: impl AsRef<SystemPath>) -> Result<ReadDirectory> {
```
Comment (MichaReiser) [inline crates/ruff_db/src/system/memory_fs.rs:256]: I'm happy to revert this pack to returning an `impl`. I had to change it, because a first version of the walker had to store the current iteration state in a `struct` which doesn't work with `impl` (Unless you use `dyn`). I introduced the named type and thought I'll leave it, because it can be useful. 

It's also nice that the type now implements `Debug`

File: crates/ruff_db/src/system.rs [hunk 2]
```diff
@@ -130,20 +139,20 @@ impl FileType {
...
     pub fn path(&self) -> &SystemPath {
         &self.path
     }
 
-    pub fn file_type(&self) -> &Result<FileType> {
-        &self.file_type
+    pub fn file_type(&self) -> FileType {
+        self.file_type
     }
 }
```
Comment (AlexWaygood) [inline crates/ruff_db/src/system.rs:157]: Heh, doesn't this just reverse the change you asked me to make in https://github.com/astral-sh/ruff/pull/12289#discussion_r1674657856? ;)

Comment (MichaReiser) [reply crates/ruff_db/src/system.rs:157]: Yeah, that's probably true. I ran into the issue that `io::Error` isn't `Clone`. I could have worked around this problem by destructing both the `path` and the `result` from `file_type`, but I'm unsure if it's worth it. 

The main advantage of `std::fs::DirEntry::file_type` is that the operation is lazy. This is different in our implementation, where we call `file_type` when returning the `DirEntry` (so it is eager). 

Now, there's a semantic difference in the new implementation. The `read_dir` call will fail when calling `file_type` fails. I don't know in what situations it is possible to know that the entry is there, but it fails to read the file type. 

I just spent some time looking into what `walker` does. It seems `walker' also eagerly reads the file type and returns an `Err` if it fails to retrieve it. 

Ideally, I think we would make the `file_type` call lazy, but that's quiet a bit more complicated because it would require holding on to `System`. 

TLDR: Not sure, but it felt like the most straightforward path and it's unclear to me what the benefit is of keeping the Result.

Comment (AlexWaygood) [reply crates/ruff_db/src/system.rs:157]: That all makes sense to me. If we're no longer storing a `Result`, though, I think I'd also make the following simplifications:

```diff
diff --git a/crates/ruff_db/src/system.rs b/crates/ruff_db/src/system.rs
index 2805137ec..5b1ce5b12 100644
--- a/crates/ruff_db/src/system.rs
+++ b/crates/ruff_db/src/system.rs
@@ -115,7 +115,7 @@ impl Metadata {
     }
 }
 
-#[derive(Copy, Clone, Eq, PartialEq, Debug, Hash)]
+#[derive(Copy, Clone, Eq, PartialEq, Debug, Hash, PartialOrd, Ord)]
 pub enum FileType {
     File,
     Directory,
@@ -136,7 +136,7 @@ impl FileType {
     }
 }
 
-#[derive(Debug)]
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord)]
 pub struct DirectoryEntry {
     path: SystemPathBuf,
     file_type: FileType,
@@ -155,9 +155,3 @@ impl DirectoryEntry {
         self.file_type
     }
 }
-
-impl PartialEq for DirectoryEntry {
-    fn eq(&self, other: &Self) -> bool {
-        self.path == other.path
-    }
-}
diff --git a/crates/ruff_db/src/system/os.rs b/crates/ruff_db/src/system/os.rs
index 2b7dd397d..99fff9449 100644
--- a/crates/ruff_db/src/system/os.rs
+++ b/crates/ruff_db/src/system/os.rs
@@ -283,7 +283,7 @@ mod tests {
             .unwrap()
             .map(Result::unwrap)
             .collect();
-        sorted_contents.sort_by(|a, b| a.path.cmp(&b.path));
+        sorted_contents.sort();
 
         let expected_contents = vec![
             DirectoryEntry::new(tempdir_path.join("a/bar.py"), FileType::File),
```

Comment (MichaReiser) [reply crates/ruff_db/src/system.rs:157]: Oh nice. I'll do that. But I'm not sure about `PartialOrd` on `FileType`. Not sure what the use case (or even the order) is.

Comment (AlexWaygood) [reply crates/ruff_db/src/system.rs:157]: > Not sure what the use case (or even the order) is.

I guess the use case is that you can easily derive `PartialOrd` on `DirectoryEntry`... but yeah, agreed that that one is questionable :)

File: crates/ruff_db/src/system/memory_fs.rs [hunk 3]
```diff
@@ -238,32 +254,33 @@ impl MemoryFileSystem {
...
+                break;
+            }
+
+            if path.parent() == Some(&normalized) {
+                collected.push(Ok(DirectoryEntry {
+                    path: SystemPathBuf::from_utf8_path_buf(path.to_owned()),
+                    file_type: entry.file_type(),
+                }));
+            }
+        }
```
Comment (AlexWaygood) [inline crates/ruff_db/src/system/memory_fs.rs:281]: nit: I don't mind returning the named type too much, but I'm not sure I see the need to change to an imperative approach here. Does doing so fix a bug? You could do this with a much smaller diff by keeping most of the code as it was:

```suggestion
        let collected = by_path
            .range(normalized.clone()..)
            .skip(1)
            .take_while(|(path, _)| path.starts_with(&normalized))
            .filter_map(|(path, entry)| {
                if path.parent()? == normalized {
                    Some(Ok(DirectoryEntry {
                        path: SystemPathBuf::from_utf8_path_buf(path.to_owned()),
                        file_type: entry.file_type(),
                    }))
                } else {
                    None
                }
            })
            .collect()
```

File: crates/ruff_db/src/system/os.rs [hunk 4]
```diff
@@ -75,11 +89,154 @@ impl System for OsSystem {
...
+                .min(12),
+        );
+
+        builder.build_parallel().run(|| {
+            let mut visitor = visitor_builder.build();
+
+            Box::new(move |entry| {
+                match entry {
+                    Ok(entry) => {
+                        // SAFETY: A system path can't be a stdin path.
```
Comment (carljm) [inline crates/ruff_db/src/system/os.rs:144]: This safety comment isn't clear to me, I assume because I'm missing some important context. How do we know that `entry` is a system path? What is a "stdin path"?

File: crates/ruff_db/src/system/os.rs [hunk 5]
```diff
@@ -75,11 +89,154 @@ impl System for OsSystem {
...
+                        let file_type = entry.file_type().unwrap();
+                        let depth = entry.depth();
+
+                        // If there's an error related to a git ignore file, warn.
+                        // Errors related to traversing the file system are reported alongside the entry.
+                        if let Some(error) = entry.error() {
+                            tracing::warn!("{error}");
+                        }
```
Comment (carljm) [inline crates/ruff_db/src/system/os.rs:152]: How do we know that this is an error related to a git ignore file?

**Review by carljm — APPROVED** [review]

File: crates/ruff_db/src/system/os.rs (hunk 5 above)
Comment (MichaReiser) [reply crates/ruff_db/src/system/os.rs:152]: This is the contract from the ignore crate.

Comment (carljm) [reply crates/ruff_db/src/system/os.rs:152]: I see; I guess it wouldn't hurt to clarify that in the comment.

File: crates/ruff_db/src/system/walk_directory.rs [hunk 6]
```diff
@@ -0,0 +1,323 @@
...
+    Loop {
+        ancestor: SystemPathBuf,
+        child: SystemPathBuf,
+    },
+
+    /// An error that occurs when doing I/O
+    Io {
+        path: Option<SystemPathBuf>,
+        err: std::io::Error,
+    },
```
[the remark this answers is not in the document]
Comment (MichaReiser) [reply crates/ruff_db/src/system/walk_directory.rs:251]: To me it depends on how the enum variants are used. Our AST used to use struct enum variants over named structs. This was very cumbersome in downstream tools because it was impossible to pass an entire `StmtFunctionDef`, because it lacked a name. That's why you still see many lint rules accepting all individual fields of their node instead of the entire node. 

Another motivation for named structs is if you want to lock-down access to individual fields because the enum fields always have the same visibility as the enum itself. 


On the other hand, using variant structs allows Rust to be more clever when it comes to layout optimizations and it's a lot less boilerplate to define the enum and to match on it (no nesting). 


I'm using struct variants here because

a) I think it's unlikely that some code would want to pass the entire variant (have it a named type) and even if, that code would at most have to accept two arguments
b) The main use case is to match on the variant and extract the fields, which is more brief with struct variants.
